pkg/packetfilter/nftables/nftables.go

```diff
@@ -0,0 +1,511 @@
+}
+
+func (p *packetFilter) chainExists(chain string) (bool, error) {
+	chainsList, err := p.nftables.List(context.TODO(), "chains")
+	if err != nil && !knftables.IsNotFound(err) {
+		return false, errors.Wrap(err, "error listing chains")
+	}
+
+	return slices.Contains(chainsList, chain), nil
+}
+
+func (p *packetFilter) AppendUnique(_ packetfilter.TableType, chain string, rule *packetfilter.Rule) error {
+	return p.insertRuleAtPosition(chain, rule, -1)
+}
+
+func (p *packetFilter) CreateIPHookChainIfNotExists(chain *packetfilter.ChainIPHook) error {
+	exists, err := p.chainExists(chain.Name)
+	if err != nil {
+		return errors.Wrap(err, "failed to check if chain exists")
+	}
+
+	if exists {
+		return nil
+	}
+
+	tx := p.nftables.NewTransaction()
+	tx.Add(&knftables.Table{
+		Comment: ptr.To("rules for submariner"),
+	})
+
+	chainPriority := iphookChainTypeToNftablesBasicPriority[chain.Type]
+	if chain.Priority == packetfilter.ChainPriorityFirst {
+		chainPriority += "-10"
+	}
+
+	nftChain := &knftables.Chain{
+		Name:     chain.Name,
+		Type:     ptr.To(iphookChainTypeToNftablesType[chain.Type]),
+		Hook:     ptr.To(iphookChainHookToNftablesHook[chain.Hook]),
+		Priority: ptr.To(chainPriority),
+	}
+	tx.Add(nftChain)
+	err = p.nftables.Run(context.TODO(), tx)
+
+	return errors.Wrap(err, "error creating IPHook Chain")
+}
+
+func (p *packetFilter) CreateChainIfNotExists(table packetfilter.TableType, chain *packetfilter.Chain) error {
+	exists, err := p.ChainExists(table, chain.Name)
+	if err != nil {
+		return errors.Wrap(err, "failed to check if chain exists")
+	}
+
+	if exists {
+		return nil
+	}
+
+	tx := p.nftables.NewTransaction()
+	tx.Add(&knftables.Table{
+		Comment: ptr.To("rules for submariner"),
+	})
+
+	tx.Add(&knftables.Chain{
+		Name: chain.Name,
+	})
+
+	err = p.nftables.Run(context.TODO(), tx)
+
+	return errors.Wrap(err, "error creating chain")
+}
+
+func (p *packetFilter) DeleteIPHookChain(chain *packetfilter.ChainIPHook) error {
+	return p.deleteChain(chain.Name)
+}
+
+func (p *packetFilter) DeleteChain(_ packetfilter.TableType, chain string) error {
+	return p.deleteChain(chain)
+}
+
+func (p *packetFilter) deleteChain(chain string) error {
+	tx := p.nftables.NewTransaction()
+	tx.Delete(&knftables.Chain{
+		Name: chain,
+	})
+
+	err := p.nftables.Run(context.TODO(), tx)
+	if knftables.IsNotFound(err) {
+		return nil
+	}
+
+	return errors.Wrap(err, "error deleting chain")
+}
+
+func (p *packetFilter) ClearChain(_ packetfilter.TableType, chain string) error {
+	tx := p.nftables.NewTransaction()
+	tx.Flush(&knftables.Chain{
+		Name: chain,
+	})
+
+	err := p.nftables.Run(context.TODO(), tx)
+	if knftables.IsNotFound(err) {
+		return nil
+	}
+
+	return errors.Wrap(err, "error clearing chain")
+}
+
+func (p *packetFilter) Delete(_ packetfilter.TableType, chain string, rule *packetfilter.Rule) error {
+	tx := p.nftables.NewTransaction()
+	tx.Delete(&knftables.Rule{
+		Chain: chain,
+		Rule:  ToRuleSpec(rule).String(),
+	})
```

Comment:
The docs for `Delete` says the `Rule` field is ignored and you must specify the `Handle`. To find the `Handle`, it seems we need to use `ListRules` and find the desired rule by its `Comment` field (see my other comment re: this).